### PR TITLE
HIVE-23235 Fix checkpointing for ORC Tables

### PR DESCRIPTION
**Pull request by aasha**

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 1]
```diff
@@ -761,6 +761,104 @@ public void testCheckPointingDataDumpFailureRegularCopy() throws Throwable {
...
+    long modifiedTimeTable1CopyFile = fs.listStatus(tablet1Path)[0].getModificationTime();
+    long modifiedTimeTable2 = fs.getFileStatus(tablet2Path).getModificationTime();
+    //Do another dump. It should only dump table t2. Modification time of table t1 should be same while t2 is greater
+    WarehouseInstance.Tuple nextDump = primary.dump(primaryDbName);
+    assertEquals(nextDump.dumpLocation, bootstrapDump.dumpLocation);
+    assertTrue(fs.exists(new Path(dumpPath, DUMP_ACKNOWLEDGEMENT.toString())));
+    //File is copied again as we are using regular copy
+    assertTrue(modifiedTimeTable1 < fs.getFileStatus(tablet1Path).getModificationTime());
+    assertTrue(modifiedTimeTable1CopyFile < fs.listStatus(tablet1Path)[0].getModificationTime());
+    assertTrue(modifiedTimeTable2 < fs.getFileStatus(tablet2Path).getModificationTime());
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:801]: In regular copy overwrite case, we are not deleting the table parent folder t2, how is this passing?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 2]
```diff
@@ -761,6 +761,104 @@ public void testCheckPointingDataDumpFailureRegularCopy() throws Throwable {
             .verifyResults(new String[]{"11", "21"});
   }
 
+  @Test
+  public void testCheckPointingDataDumpFailureORCTableRegularCopy() throws Throwable {
+    WarehouseInstance.Tuple bootstrapDump = primary.run("use " + primaryDbName)
+            .run("CREATE TABLE t1(a int) clustered by (a) into 2 buckets" +
+                    " stored as orc TBLPROPERTIES ('transactional'='true')")
+            .run("CREATE TABLE t2(a string) STORED AS TEXTFILE")
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:769]: Add one empty table too.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java [hunk 3]
```diff
@@ -761,6 +761,104 @@ public void testCheckPointingDataDumpFailureRegularCopy() throws Throwable {
...
+    List<String> dumpClause = Arrays.asList(
+            "'" + HiveConf.ConfVars.HIVE_EXEC_COPYFILE_MAXSIZE.varname + "'='1'",
+            "'" + HiveConf.ConfVars.HIVE_IN_TEST.varname + "'='false'",
+            "'" + HiveConf.ConfVars.HIVE_EXEC_COPYFILE_MAXNUMFILES.varname + "'='0'",
+            "'" + HiveConf.ConfVars.HIVE_DISTCP_DOAS_USER.varname + "'='"
+                    + UserGroupInformation.getCurrentUser().getUserName() + "'");
+    WarehouseInstance.Tuple bootstrapDump = primary.run("use " + primaryDbName)
+            .run("CREATE TABLE t1(a int) clustered by (a) into 2 buckets" +
+                    " stored as orc TBLPROPERTIES ('transactional'='true')")
+            .run("CREATE TABLE t2(a string) STORED AS TEXTFILE")
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosAcidTables.java:822]: t2 should be ORC table as the tests are mainly dealing with overwrite to it.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java [hunk 4]
```diff
@@ -383,10 +385,17 @@ private void doRegularCopyOnce(FileSystem sourceFs, List<Path> srcList,
         throw new IOException(e);
       }
     } else {
+      deleteSubDirs(destinationFs, destination);
       FileUtil.copy(sourceFs, paths, destinationFs, destination, false, true, hiveConf);
     }
   }
 
+  private void deleteSubDirs(FileSystem destinationFs, Path destination) throws IOException {
+    for (FileStatus status : destinationFs.listStatus(destination)) {
```
[the remark this answers is not in the document]
Comment (aasha) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/repl/CopyUtils.java:394]: We need the destination folder to be present. Thats an extra FS call if we delete first and recreate again.

**Pull request closed without merging**